backend/event/design.md

```diff
@@ -49,6 +49,89 @@ The producer will start when the service start, each micro-service constains an
+   */
+  private String customerId;
+
+  /**
+   * anonymous id.
+   */
+  private String anonymousId;
+```
+### 3.6.2 Define the topics
+Different broker has different way to create topic.
+For now, we use google's pub/sub, so fellow this doc: [google pub/sub](https://cloud.google.com/pubsub/docs/)
+### 3.6.3 Define the event entity
+When some thing happened, then the somain service produce a event and save it to database.
+Each micro service only have one event entity(Save all event to one table), or have many different event entity. Use one event entity is suggested, because we use the common `message` to publish our events.
+For `customer-authentication` service, the event entity is:
+```Java
+public class EventMessage {
+
+  private String id;
+
+  private long createdAt;
+
+  private String sequenceNumber;
+
+  private long expire;
+
+  private Integer version;
+
+  private String payload;
+
+  // Which topic this event blong to.
+  private String topic;
+
+  // Event status, CREATED if waiting for publish, PENDING is being publish.
+  private EventStatus status;
+}
+```
+### 3.6.4 Define the event creator
+For some business actions, they may produce some event. so create the event when the action triggered.
+**Make sure that the business action and the action for create event are in one transaction.**
+Create event action is create an event and save it to database.
+
+### 3.6.5 Publish the event to event bus
+Event publisher is a service that circulate running. Since we used springboot, so we can use the `Scheduled Job` for this service.
+#### 3.6.5.1 Create the scheduled job
+Scheduled job call the publisher to publish the event to event bus.
+In springboot, we can use `Scheduled Job`:
+- first, add `@EnableScheduling` to the `Application` class
+- second, create a service, and add `@Scheduled(fixedRate = 200)` to one of the functions in this service. fixedRate means the function will executes each 200 ms. For example:
+```java
+@Service
+public class EventPublisher {
+  @Scheduled(fixedRate = 200)
+  public void executor() {
+    // do the publish action
+  }
+}
+```
+#### 3.6.5.2 Create the publisher client with `message-client` lib
+Message-client lib is the lib we used to publish message to event bus.
+- first, inport the dependency, see README of `message-client`
+- second, create the producer.
+For now, we use google's pub/sub, so use this to create the producer client:
+```java
+Producer producer = DefaultProducerFactory.createGoogleProducer(`projectId`, `topicName`);
+```
+As we may got several topic, so we need to create the same number of producer.
+#### 3.6.5.3 publish the event to event bus
+publish the event to event bus, then the producer's work is done.
+- first, fetch the events from db, and change their status to `PENDING`. or fetch those event whose status is `PENDING`, but created long time ago(maybe one minute ago).
+- second, create messages with those events.
+- third, publish those messages to event bus.
+
```

Comment:
in 3.5 Workflow, there is a setp "After publish success, delete the event from db", but i can not find the detail description.